### PR TITLE
Druid filters in configured timezone

### PR DESCRIPTION
**Pull request by moumny**
Related to #3386 .
Filters in druid (intervals) are now made in the report timezone configured instead of UTC.

### CODE REVIEW

**Comment by salsakran** [comment]
Awesome.

Could you add some tests that show the bug so we can make sure there are no regressions in the future?

File: src/metabase/util.clj [hunk 1]
```diff
@@ -200,40 +200,42 @@
 
      (date-trunc :month).
      ;; -> #inst \"2015-11-01T00:00:00\""
-  (^java.sql.Timestamp [unit]
+  (^java.sql.Timestamp [unit timezone-id]
    (date-trunc unit (System/currentTimeMillis)))
-  (^java.sql.Timestamp [unit date]
-   (let [trunc-with-format (fn trunc-with-format
+  (^java.sql.Timestamp [unit date timezone-id]
```
Comment (camsaul) [inline src/metabase/util.clj]: This is definitely going to break things, there's lots of places in the code where we user this function and they're not passing the extra timezone arg

**Comment by camsaul** [comment]
@moumny, the changes you made in `metabase.util` will definitely break some of our other drivers. `date-trunc` is used all over the place. Changing the arity of functions like that will break all of the other code that calls those functions.

**Comment by camsaul** [comment]
The current iteration of `date-trunc` is a 1 or 2 arity function, so when you change it to 2 or 3 all the places that were calling the 1-arity version will throw an `ArityException` and all the places that call the 2-arity version will be passing the wrong arguments.

If you need to pass a third argument then a better solution would be to add a third arity of the function

**Comment by moumny** [comment]
Sorry should have put WIP in the title.
The functions are only used in druid and mongo driver.
I decided to update mongo driver to make it explicit that the timezone is UTC.

We should probably add some test in timeseries testsuite that take into account the report timezone and test the timestamp filters. Would need bit of help for that ^^. (especially that mongo for example filters still in UTC i guess)

**Comment by moumny** [comment]
✅ mongodb tests are passing :)

**Comment by camsaul** [comment]
@moumny this is looking good but as I mentioned above

> If you need to pass a third argument then a better solution would be to add a third arity of the function

i.e. if "UTC" is basically the default `timezone-id` argument for things like `relative-date` and `date-trunc` I'd rather keep the existing arities and have those pass "UTC" as a default arg, and then add a new arity that takes `timezone-id`. For example, I'd prefer `date-extract` look like this after your changes:

```clojure
(defn date-extract
  ([unit]
   (date-extract unit (System/currentTimeMillis)))
  ([unit date]
   (date-extract unit date "UTC"))
  ([unit date timezone-id])
   ...))
```

That way the Mongo driver wouldn't have to be changed at all. It looks like you did this with `->Calendar` already

File: src/metabase/driver/druid/query_processor.clj [hunk 2]
```diff
@@ -229,6 +231,7 @@
     :year            (extract:timeFormat "yyyy")))
 
 (defn- unit->granularity [unit]
+  (let [timeZone (get-timezone-id)]
```
Comment (camsaul) [inline src/metabase/driver/druid/query_processor.clj]: is this being used?

File: src/metabase/util.clj [hunk 3]
```diff
@@ -200,40 +203,46 @@
...
+   (let [tz                (t/time-zone-for-id timezone-id)
+         with-zone-fmt     (fn [format-string] (time/with-zone (time/formatter format-string) tz))
+         trunc-with-floor  (fn [date amount-ms]
+                             (let [orig-ms    (.getTime (->Timestamp date))
+                                   trunced-ms (-> orig-ms (/ amount-ms) (math/floor) (* amount-ms))]
+                                (new java.sql.Timestamp trunced-ms)))
```
Comment (camsaul) [inline src/metabase/util.clj]: https://github.com/bbatsov/clojure-style-guide#sugared-java-interop

Comment (camsaul) [inline src/metabase/util.clj]: Actually you should probably just do `(->Timestamp trunced-ms)`

File: test/metabase/util_test.clj [hunk 4]
```diff
@@ -6,45 +6,85 @@
...
+(expect #inst "2006-01-01+09:00" (date-trunc :month   saturday-the-31st   "Asia/Tokyo"))
+(expect #inst "2006-01-01+09:00" (date-trunc :quarter saturday-the-31st   "Asia/Tokyo"))
+
+(expect #inst "2005-12-31T19:05" (date-trunc :minute  saturday-the-31st   "US/Pacific"))
+(expect #inst "2005-12-31T19:00" (date-trunc :hour    saturday-the-31st   "US/Pacific"))
+(expect #inst "2005-12-31-08:00" (date-trunc :day     saturday-the-31st   "US/Pacific"))
+(expect #inst "2005-12-25-08:00" (date-trunc :week    saturday-the-31st   "US/Pacific"))
+(expect #inst "2005-12-25-08:00" (date-trunc :week    sunday-the-1st      "US/Pacific"))
+(expect #inst "2005-12-01-08:00" (date-trunc :month   saturday-the-31st   "US/Pacific"))
+(expect #inst "2005-10-01-08:00" (date-trunc :quarter saturday-the-31st   "US/Pacific"))
```
Comment (camsaul) [inline test/metabase/util_test.clj]: 👍  on the extra tests

**Comment by moumny** [comment]
OK I changed the arity to be consistent with the rest. The idea was to avoid defaults to UTC to avoid mistakes. Its convenient to call one of the first UTC methods in druid driver and its harder to spot than (date-op  date "UTC") :) 

Anyway let me know if you see something else. 👁

**Comment by camsaul** [comment]
@moumny, this looks good 👍 . I'll run it on CI in a bit and if all the tests pass then we can merge this

Thanks for your contribution 😍

**Comment by camsaul** [comment]
@moumny I had to make a couple last-minute tweaks to this so we can ship it as part of our 0.20 release. I opened a new PR for it, #3459, so I'm going to go ahead and close this one out.

I'll merge your fixes in as soon as the tests pass on CI. Thanks again for your contribution! 🎉

**Pull request closed without merging**